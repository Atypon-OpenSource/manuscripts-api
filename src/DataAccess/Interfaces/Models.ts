/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Invitation,
  Project,
  Library,
  LibraryCollection,
  ContainerInvitation,
  Collaboration,
  UserProfile,
  ContainerRequest,
  Submission,
  ManuscriptNote,
  ExternalFile,
  Manuscript,
  Correction,
  ManuscriptTemplate,
  Snapshot,
} from '@manuscripts/manuscripts-json-schema'

export type ProjectLike = Pick<Project, Exclude<keyof Project, 'createdAt' | 'updatedAt'>>

export type LibraryLike = Pick<Library, Exclude<keyof Library, 'createdAt' | 'updatedAt'>>

export type LibraryCollectionLike = Pick<
  LibraryCollection,
  Exclude<keyof LibraryCollection, 'createdAt' | 'updatedAt'>
>

export type InvitationLike = Pick<Invitation, Exclude<keyof Invitation, 'createdAt' | 'updatedAt'>>

export type UserProfileLike = Pick<
  UserProfile,
  Exclude<keyof UserProfile, 'createdAt' | 'updatedAt'>
>

export type ContainerInvitationLike = Pick<
  ContainerInvitation,
  Exclude<keyof ContainerInvitation, 'createdAt' | 'updatedAt'>
> & { expiry: number }

export type CollaborationLike = Pick<
  Collaboration,
  Exclude<keyof Collaboration, 'createdAt' | 'updatedAt'>
>

export type ContainerRequestLike = Pick<
  ContainerRequest,
  Exclude<keyof ContainerRequest, 'createdAt' | 'updatedAt'>
>

export type SubmissionLike = Pick<Submission, Exclude<keyof Submission, 'createdAt' | 'updatedAt'>>

export type ManuscriptNoteLike = Pick<
  ManuscriptNote,
  Exclude<keyof ManuscriptNote, 'createdAt' | 'updatedAt'>
>

export type ManuscriptLike = Pick<Manuscript, Exclude<keyof Manuscript, 'createdAt' | 'updatedAt'>>

export type ExternalFileLike = Pick<
  ExternalFile,
  Exclude<keyof ExternalFile, 'createdAt' | 'updatedAt'>
>

export type CorrectionLike = Pick<Correction, Exclude<keyof Correction, 'createdAt' | 'updatedAt'>>

export type ManuscriptTemplateLike = Pick<
  ManuscriptTemplate,
  Exclude<keyof ManuscriptTemplate, 'createdAt' | 'updatedAt'>
>

export type SnapshotLike = Pick<Snapshot, Exclude<keyof Snapshot, 'createdAt' | 'updatedAt'>>

export type PrismaModelLike = {
  id: string
  data: any
}
