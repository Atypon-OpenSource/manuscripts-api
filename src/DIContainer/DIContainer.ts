/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { config } from '../Config/Config'
import { BucketKey } from '../Config/ConfigurationTypes'
import applyMiddleware from '../DataAccess/applyMiddleware'
import { ClientApplicationRepository } from '../DataAccess/ClientApplicationRepository/ClientApplicationRepository'
import { CollaborationsRepository } from '../DataAccess/CollaborationsRepository/CollaborationsRepository'
import { ContainerInvitationRepository } from '../DataAccess/ContainerInvitationRepository/ContainerInvitationRepository'
import { ContainerRequestRepository } from '../DataAccess/ContainerRequestRepository/ContainerRequestRepository'
import { IClientApplicationRepository } from '../DataAccess/Interfaces/IClientApplicationRepository'
import { ICollaborationsRepository } from '../DataAccess/Interfaces/ICollaborationsRepository'
import { IInvitationTokenRepository } from '../DataAccess/Interfaces/IInvitationTokenRepository'
import { IManuscriptRepository } from '../DataAccess/Interfaces/IManuscriptRepository'
import { RepositoryLike, SGRepositoryLike } from '../DataAccess/Interfaces/IndexedRepository'
import { ISingleUseTokenRepository } from '../DataAccess/Interfaces/ISingleUseTokenRepository'
import { IUserEmailRepository } from '../DataAccess/Interfaces/IUserEmailRepository'
import { IUserEventRepository } from '../DataAccess/Interfaces/IUserEventRepository'
import { IUserRepository } from '../DataAccess/Interfaces/IUserRepository'
import { IUserStatusRepository } from '../DataAccess/Interfaces/IUserStatusRepository'
import { IUserTokenRepository } from '../DataAccess/Interfaces/IUserTokenRepository'
import { InvitationRepository } from '../DataAccess/InvitationRepository/InvitationRepository'
import { InvitationTokenRepository } from '../DataAccess/InvitationTokenRepository/InvitationTokenRepository'
import { ManuscriptNoteRepository } from '../DataAccess/ManuscriptNoteRepository/ManuscriptNoteRepository'
import { ManuscriptRepository } from '../DataAccess/ManuscriptRepository/ManuscriptRepository'
import { ProjectRepository } from '../DataAccess/ProjectRepository/ProjectRepository'
import { SingleUseTokenRepository } from '../DataAccess/SingleUseTokenRepository/SingleUseTokenRepository'
import { SQLDatabase } from '../DataAccess/SQLDatabase'
import { TemplateRepository } from '../DataAccess/TemplateRepository/TemplateRepository'
import { UserCollaboratorRepository } from '../DataAccess/UserCollaboratorRepository/UserCollaboratorRepository'
import { UserEmailRepository } from '../DataAccess/UserEmailRepository/UserEmailRepository'
import { UserEventRepository } from '../DataAccess/UserEventRepository/UserEventRepository'
import { UserProfileRepository } from '../DataAccess/UserProfileRepository/UserProfileRepository'
import { UserRepository } from '../DataAccess/UserRepository/UserRepository'
import { UserStatusRepository } from '../DataAccess/UserStatusRepository/UserStatusRepository'
import { UserTokenRepository } from '../DataAccess/UserTokenRepository/UserTokenRepository'
import { AuthService } from '../DomainServices/Auth/AuthService'
import { IAuthService } from '../DomainServices/Auth/IAuthService'
import { ContainerService } from '../DomainServices/Container/ContainerService'
import { ContainerRequestService } from '../DomainServices/ContainerRequest/ContainerRequestService'
import { IContainerRequestService } from '../DomainServices/ContainerRequest/IContainerRequestService'
import { EmailService } from '../DomainServices/Email/EmailService'
import { ExpirationService } from '../DomainServices/Expiration/ExpirationService'
import { ContainerInvitationService } from '../DomainServices/Invitation/ContainerInvitationService'
import { IContainerInvitationService } from '../DomainServices/Invitation/IContainerInvitationService'
import { InvitationService } from '../DomainServices/Invitation/InvitationService'
import { IPressroomService } from '../DomainServices/Pressroom/IPressroomService'
import { PressroomService } from '../DomainServices/Pressroom/PressroomService'
import { ProjectService } from '../DomainServices/ProjectService'
import { IQuarterbackService } from '../DomainServices/Quarterback/IQuarterbackService'
import { QuarterbackService } from '../DomainServices/Quarterback/QuarterbackService'
import { IUserRegistrationService } from '../DomainServices/Registration/IUserRegistrationService'
import { UserRegistrationService } from '../DomainServices/Registration/UserRegistrationService'
import { ISGService } from '../DomainServices/SG/ISGService'
import { SGService } from '../DomainServices/SG/SGService'
import { ISyncService } from '../DomainServices/Sync/ISyncService'
import { SyncService } from '../DomainServices/Sync/SyncService'
import { IUserService } from '../DomainServices/User/IUserService'
import { UserService } from '../DomainServices/User/UserService'
import { UserActivityTrackingService } from '../DomainServices/UserActivity/UserActivityTrackingService'
import { IServer } from '../Server/IServer'
import { Server } from '../Server/Server'

export class UninitializedContainerError extends Error {
  constructor() {
    super()
    Object.setPrototypeOf(this, new.target.prototype)
  }
}

export class ContainerReinitializationError extends Error {
  constructor() {
    super()
    Object.setPrototypeOf(this, new.target.prototype)
  }
}

export class DIContainer {
  private static _sharedContainer: DIContainer | null = null

  static get sharedContainer(): DIContainer {
    if (DIContainer._sharedContainer === null) {
      throw new UninitializedContainerError()
    } else {
      return DIContainer._sharedContainer
    }
  }

  readonly server: IServer
  readonly userRepository: IUserRepository
  readonly userTokenRepository: IUserTokenRepository
  readonly userEmailRepository: IUserEmailRepository
  readonly singleUseTokenRepository: ISingleUseTokenRepository
  readonly applicationRepository: IClientApplicationRepository
  readonly collaborationsRepository: ICollaborationsRepository
  readonly emailService: EmailService
  readonly expirationService: ExpirationService
  readonly syncService: ISyncService
  readonly sgService: ISGService
  readonly authService: IAuthService
  readonly userRegistrationService: IUserRegistrationService
  readonly activityTrackingService: UserActivityTrackingService
  readonly userStatusRepository: IUserStatusRepository
  readonly userEventRepository: IUserEventRepository
  readonly userService: IUserService
  readonly invitationRepository: InvitationRepository
  readonly containerInvitationRepository: ContainerInvitationRepository
  readonly invitationTokenRepository: IInvitationTokenRepository
  readonly containerInvitationService: IContainerInvitationService
  readonly invitationService: InvitationService
  readonly projectRepository: ProjectRepository
  readonly userProfileRepository: UserProfileRepository
  readonly userCollaboratorRepository: UserCollaboratorRepository
  readonly containerService: ContainerService
  readonly projectService: ProjectService
  readonly containerRequestService: IContainerRequestService
  readonly containerRequestRepository: ContainerRequestRepository
  readonly pressroomService: IPressroomService
  readonly quarterback: IQuarterbackService
  readonly manuscriptRepository: IManuscriptRepository
  readonly manuscriptNotesRepository: ManuscriptNoteRepository
  readonly templateRepository: TemplateRepository

  /**
   * WARNING: internal method.
   *
   * We need to await promises (SQLDatabase initialization), which cannot be done
   * in a constructor.
   *
   * This constructor should only be called by the DIContainer.init() method.
   */
  constructor(
    readonly userBucket: SQLDatabase,
    readonly dataBucket: SQLDatabase,
    readonly enableActivityTracking: boolean
  ) {
    this.userCollaboratorRepository = new UserCollaboratorRepository(
      BucketKey.User,
      this.userBucket
    )
    this.applicationRepository = new ClientApplicationRepository(this.userBucket)
    this.server = new Server(this.userBucket)
    this.userRepository = new UserRepository(this.userBucket)
    this.userTokenRepository = new UserTokenRepository(this.userBucket)
    this.userEmailRepository = new UserEmailRepository(this.userBucket)
    this.singleUseTokenRepository = new SingleUseTokenRepository(this.userBucket)
    this.invitationTokenRepository = new InvitationTokenRepository(this.userBucket)
    this.collaborationsRepository = new CollaborationsRepository(BucketKey.Project, this.dataBucket)
    this.emailService = new EmailService(config.email, {})
    this.userEventRepository = new UserEventRepository(this.userBucket)
    this.activityTrackingService = new UserActivityTrackingService(
      this.userEventRepository,
      this.enableActivityTracking
    )
    this.userStatusRepository = new UserStatusRepository(this.userBucket)
    this.userProfileRepository = new UserProfileRepository(BucketKey.Project, this.dataBucket)
    this.syncService = new SyncService(this.userStatusRepository, this.userProfileRepository)
    this.sgService = new SGService()
    this.userRegistrationService = new UserRegistrationService(
      this.userRepository,
      this.userEmailRepository,
      this.emailService,
      this.singleUseTokenRepository,
      this.activityTrackingService,
      this.userStatusRepository,
      this.syncService
    )
    this.projectRepository = new ProjectRepository(BucketKey.Project, this.dataBucket)
    this.invitationRepository = new InvitationRepository(BucketKey.Project, this.dataBucket)
    this.containerInvitationRepository = new ContainerInvitationRepository(
      BucketKey.Project,
      this.dataBucket
    )
    this.containerRequestRepository = new ContainerRequestRepository(
      BucketKey.Project,
      this.dataBucket
    )
    this.manuscriptRepository = new ManuscriptRepository(BucketKey.Project, this.dataBucket)
    this.manuscriptNotesRepository = new ManuscriptNoteRepository(
      BucketKey.Project,
      this.dataBucket
    )
    this.templateRepository = new TemplateRepository(BucketKey.Project, this.dataBucket)
    this.userService = new UserService(
      this.userRepository,
      this.singleUseTokenRepository,
      this.activityTrackingService,
      this.userStatusRepository,
      this.userTokenRepository,
      this.invitationRepository,
      this.containerInvitationRepository,
      this.containerRequestRepository,
      this.emailService,
      this.syncService,
      this.userProfileRepository,
      this.projectRepository,
      this.userCollaboratorRepository
    )
    this.containerService = new ContainerService(
      this.userRepository,
      this.userService,
      this.activityTrackingService,
      this.userStatusRepository,
      this.projectRepository,
      this.containerInvitationRepository,
      this.emailService,
      this.manuscriptRepository,
      this.manuscriptNotesRepository,
      this.templateRepository
    )
    this.projectService = new ProjectService(
      this.projectRepository,
      this.manuscriptRepository,
      this.templateRepository,
      this.userRepository
    )
    this.containerInvitationService = new ContainerInvitationService(
      this.userRepository,
      this.userProfileRepository,
      this.emailService,
      this.containerService,
      this.containerInvitationRepository,
      this.invitationTokenRepository,
      this.activityTrackingService
    )
    this.invitationService = new InvitationService(
      this.userRepository,
      this.userProfileRepository,
      this.emailService,
      this.invitationRepository,
      this.collaborationsRepository,
      this.activityTrackingService,
      this.userRegistrationService
    )
    this.containerRequestService = new ContainerRequestService(
      this.containerRequestRepository,
      this.userProfileRepository,
      this.userRepository,
      this.containerService,
      this.emailService
    )
    this.authService = new AuthService(
      this.userRepository,
      this.userTokenRepository,
      this.userProfileRepository,
      this.activityTrackingService,
      this.syncService,
      this.userStatusRepository
    )
    this.expirationService = new ExpirationService(
      this.userEventRepository,
      this.userTokenRepository,
      this.invitationRepository,
      this.invitationTokenRepository,
      this.containerInvitationRepository
    )
    this.pressroomService = new PressroomService(config.pressroom.baseurl, config.pressroom.apiKey)
    this.quarterback = new QuarterbackService(config.quarterback.baseurl, config.quarterback.apiKey)
  }

  /**
   * Initializes the container
   *
   * This method creates a SQLDatabase instance and passes it to the
   * services/repositories.
   *
   * This method should be called once at app startup.
   *
   * The server is then retrieved from the container and bootstrapped.
   */
  static async init(enableActivityTracking = false) {
    if (DIContainer._sharedContainer !== null) {
      throw new ContainerReinitializationError()
    }
    const userBucket = new SQLDatabase(config.DB, BucketKey.User)

    // no loading of database models needed from this bucket (no Ottoman models are mapped there).
    const dataBucket = new SQLDatabase(config.DB, BucketKey.Project)

    // do NOT parallelise these. Deferred PRIMARY index creation appears buggy in CB.
    await userBucket.loadDatabaseModels()
    await dataBucket.loadDatabaseModels()

    applyMiddleware()

    DIContainer._sharedContainer = new DIContainer(userBucket, dataBucket, enableActivityTracking)

    await DIContainer._sharedContainer.applicationRepository.ensureApplicationsExist(
      (config.apps && config.apps.knownClientApplications) || []
    )

    return DIContainer._sharedContainer
  }

  public bucketForKey(bucketKey: BucketKey): SQLDatabase {
    switch (bucketKey) {
      case BucketKey.User:
        return this.userBucket
      case BucketKey.Project:
        return this.dataBucket
      default:
        return this.dataBucket
    }
  }

  public static isRepositoryLike(obj: any): obj is RepositoryLike {
    return obj && obj.create != null && obj.documentType != null
  }

  public static isSGRepositoryLike(obj: any): obj is SGRepositoryLike {
    return obj && obj.create != null && obj.objectType != null
  }

  public get repositories(): RepositoryLike[] {
    const repositories = (Object as any).values(this).filter(DIContainer.isRepositoryLike)
    const repositorySet: Set<RepositoryLike> = new Set(repositories)
    return Array.from(repositorySet)
  }

  public get gatewayRepositories(): SGRepositoryLike[] {
    const repositories = (Object as any).values(this).filter(DIContainer.isSGRepositoryLike)
    const repositorySet: Set<SGRepositoryLike> = new Set(repositories)
    return Array.from(repositorySet)
  }
}
