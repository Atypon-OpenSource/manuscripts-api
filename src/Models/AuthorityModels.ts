/*!
 * Â© 2024 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Prisma } from '@prisma/client'
import { Step } from 'prosemirror-transform'

export type ReceiveSteps = {
  steps: Prisma.JsonObject[]
  clientID: number
  version: number
}

export type History = {
  steps: Step[] | Prisma.JsonValue[]
  clientIDs: number[]
  version: number
  doc?: Prisma.JsonValue
}

export type DocumentHistory = History & { doc: Prisma.JsonValue | undefined }
export interface ModifiedStep extends Prisma.JsonObject {
  clientID: string
}
