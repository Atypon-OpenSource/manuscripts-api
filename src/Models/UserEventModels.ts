/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export enum UserActivityEventType {
  DeleteAccount = 'DeleteAccount',
  EmailVerified = 'EmailVerified',
  Logout = 'Logout',
  RefreshSyncSession = 'RefreshSyncSession',
  Registration = 'Registration',
  ResetPassword = 'ResetPassword',
  SendPasswordResetEmail = 'SendPasswordResetEmail',
  SendInvitation = 'SendInvitation',
  StatusNotFound = 'StatusNotFound',
  SuccessfulLogin = 'SuccessfulLogin',
  RequestEmailVerification = 'RequestEmailVerification',
  ProjectCreated = 'ProjectCreated',
}

export interface UserActivityEvent {
  _id?: string
  userId: string
  createdAt: Date
  deviceId: string | null
  appId: string | null
  eventType: UserActivityEventType
}
