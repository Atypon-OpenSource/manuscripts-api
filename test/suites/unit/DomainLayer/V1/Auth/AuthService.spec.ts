/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '../../../../../../test/utilities/configMock'
import '../../../../../../test/utilities/dbMock'

import { describe } from 'jest-circus'

import { DIContainer } from '../../../../../../src/DIContainer/DIContainer'
import {
  AuthService,
  MAX_NUMBER_OF_LOGIN_ATTEMPTS,
} from '../../../../../../src/DomainServices/Auth/AuthService'
import {
  AccountNotFoundError,
  InvalidCredentialsError,
  InvalidPasswordError,
  MissingUserStatusError,
  UserBlockedError,
  UserNotVerifiedError,
  ValidationError,
} from '../../../../../../src/Errors'
import { UserActivityEventType } from '../../../../../../src/Models/UserEventModels'
import { ServerToServerAuthCredentials } from '../../../../../../src/Models/UserModels'
import {
  userWithValidCredentials,
  validJWTToken,
  validUserStatus,
  validUserToken,
} from '../../../../../data/fixtures/authServiceUser'
import {
  invalidCredentials,
  invalidPasswordCredentials,
  validCredentials,
} from '../../../../../data/fixtures/credentials'
import { validBody } from '../../../../../data/fixtures/credentialsRequestPayload'
import { validUser } from '../../../../../data/fixtures/userServiceUser'
import { TEST_TIMEOUT } from '../../../../../utilities/testSetup'

jest.mock('../../../../../../src/DomainServices/Sync/SyncService', () => {
  return {
    GATEWAY_BUCKETS: ['data', 'shared'],
    SyncService: jest.fn(() => ({
      createGatewaySessions: jest.fn(() => ['valid-sync-session']),
      getOrCreateUserStatus: jest.fn(),
      createUserProfile: jest.fn(),
      createGatewayAdministrator: jest.fn(),
      removeGatewaySessions: jest.fn(),
      removeAllGatewaySessions: jest.fn(),
    })),
  }
})

jest.setTimeout(TEST_TIMEOUT)

beforeEach(() => {
  ;(DIContainer as any)._sharedContainer = null
  return DIContainer.init(true)
})

describe('AuthService - Login', () => {
  test('should fail if email does not exist', () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: () => Promise.resolve(null),
    }

    return expect(authService.login(invalidCredentials)).rejects.toThrow(AccountNotFoundError)
  })

  test("should fail if user status doesn't exists", () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: async () => Promise.resolve(userWithValidCredentials),
    }
    authService.userStatusRepository = {
      statusForUserId: async () => Promise.resolve(null),
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }

    return expect(authService.login(invalidPasswordCredentials)).rejects.toThrow(
      MissingUserStatusError
    )
  })

  test('should fail if user is blocked', () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: async () => Promise.resolve(userWithValidCredentials),
    }
    authService.userStatusRepository = {
      statusForUserId: async () =>
        Promise.resolve({
          blockUntil: new Date().getTime() + 2000,
          isVerified: true,
        }),
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }

    return expect(authService.login(invalidPasswordCredentials)).rejects.toThrow(UserBlockedError)
  })

  test('should fail if password is wrong and block user', () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: async () => Promise.resolve(userWithValidCredentials),
    }
    authService.userStatusRepository = {
      statusForUserId: async () =>
        Promise.resolve({
          isBlocked: true,
          blockUntil: new Date().getTime() - 5000,
          password: '123',
          isVerified: true,
        }),
      failedLoginCount: async () => Promise.resolve(MAX_NUMBER_OF_LOGIN_ATTEMPTS),
      patch: async () => Promise.resolve(),
      fullyQualifiedId: (id: string) => `UserStatus|${id}`,
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }

    return expect(authService.login(invalidPasswordCredentials)).rejects.toThrow(
      InvalidPasswordError
    )
  })

  test('should fail if password is wrong', async () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: () => Promise.resolve(userWithValidCredentials),
    }
    authService.userStatusRepository = {
      statusForUserId: async () =>
        Promise.resolve({
          blockUntil: new Date().getTime() - 5000,
          password: '123',
          isVerified: true,
        }),
      failedLoginCount: async () => Promise.resolve(0),
      patch: async () => Promise.resolve(),
      fullyQualifiedId: (id: string) => `UserStatus|${id}`,
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }

    return expect(authService.login(invalidPasswordCredentials)).rejects.toThrow(
      InvalidPasswordError
    )
  })

  test('should fail if user is not verified', () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: () => Promise.resolve(userWithValidCredentials),
    }
    authService.userStatusRepository = {
      statusForUserId: async () =>
        Promise.resolve({
          isBlocked: false,
          blockUntil: null,
          password: 'the-password',
          isVerified: false,
        }),
      failedLoginCount: async () => Promise.resolve(MAX_NUMBER_OF_LOGIN_ATTEMPTS),
      patch: async () => Promise.resolve(),
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }

    return expect(authService.login(validCredentials)).rejects.toThrow(UserNotVerifiedError)
  })

  test('should log user in and generates new user token', async () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: () => Promise.resolve(userWithValidCredentials),
    }
    authService.userStatusRepository = {
      statusForUserId: async () => Promise.resolve(validUserStatus),
      failedLoginCount: async () => Promise.resolve(0),
      patch: async () => Promise.resolve(),
      fullyQualifiedId: (id: string) => `UserStatus|${id}`,
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }
    authService.userTokenRepository = {
      getOne: async () => Promise.resolve(null),
      create: async () => Promise.resolve(null),
      touch: async () => Promise.resolve(null),
      fullyQualifiedId: (id: string) => `User|${id}`,
    }

    authService.userProfileRepository = {
      getById: async () => Promise.resolve({}),
    }

    const user = await authService.login(validCredentials)

    expect(user.token).toBeDefined()
    delete user.token

    expect(user).toMatchSnapshot()
  })

  test('should log user in and updates existing token', async () => {
    const authService: any = DIContainer.sharedContainer.authService

    authService.userRepository = {
      getOne: () => Promise.resolve(userWithValidCredentials),
    }

    authService.userTokenRepository = {
      getOne: () => Promise.resolve(validUserToken),
      create: () => Promise.resolve(null),
      patch: () => Promise.resolve(null),
      touch: () => Promise.resolve(null),
      fullyQualifiedId: (id: string) => `User|${id}`,
    }
    authService.userStatusRepository = {
      statusForUserId: async () => Promise.resolve(validUserStatus),
      failedLoginCount: async () => Promise.resolve(0),
      patch: async () => Promise.resolve(),
      fullyQualifiedId: (id: string) => `UserStatus|${id}`,
    }
    authService.userEventRepository = {
      create: async () => Promise.resolve(null),
    }

    authService.userProfileRepository = {
      getById: async () => Promise.resolve({}),
    }

    const user = await authService.login(validCredentials)

    expect(user.token).toBeDefined()

    delete user.token
    expect(user).toMatchSnapshot()
  })
})

describe('AuthService - createEvent', () => {
  test('should fail if error occurred', () => {
    const activityTrackingService: any = DIContainer.sharedContainer.activityTrackingService
    activityTrackingService.userEventRepository = {
      create: function () {
        return new Promise((_resolve, reject) => reject(new Error('User tracking derp')))
      },
    }

    return expect(
      activityTrackingService.createEvent('123', UserActivityEventType.EmailVerified, null, null)
    ).rejects.toThrow()
  })

  test('should create event successfully', async () => {
    const activityTrackingService: any = DIContainer.sharedContainer.activityTrackingService
    activityTrackingService.userEventRepository = {
      create: () => {
        return new Promise((resolve, _reject) =>
          resolve({
            userId: '123',
            userActivity: UserActivityEventType.EmailVerified,
            deviceId: null,
          })
        )
      },
    }

    await activityTrackingService.awaitCreation()

    const activity = await activityTrackingService.createEvent(
      '123',
      UserActivityEventType.EmailVerified,
      null
    )

    expect(activity).toEqual({
      userId: '123',
      userActivity: UserActivityEventType.EmailVerified,
      deviceId: null,
    })
  })
})

describe('AuthService - ensureUserStatusExists', () => {
  test('should call getOrCreateUserStatus if gateway account does not exists', async () => {
    const authService: any = DIContainer.sharedContainer.authService
    authService.syncService = {
      getOrCreateUserStatus: jest.fn(),
    }
    await authService.ensureUserStatusExists('userId')
    expect(authService.syncService.getOrCreateUserStatus).toHaveBeenCalled()
  })
})

describe('AuthService - isBearerHeaderValue', () => {
  test('should return false if the value is not a Bearer value', () => {
    expect(AuthService.isBearerHeaderValue('foo')).toBeFalsy()
  })

  test('should return false if the value is not a string', () => {
    expect(AuthService.isBearerHeaderValue(123 as any)).toBeFalsy()
  })

  test('should return true if the value is Bearer and string', () => {
    expect(AuthService.isBearerHeaderValue('Bearer foo')).toBeTruthy()
  })
})

describe('AuthService - ensureValidAuthorizationBearer', () => {
  test('should fail if the value is not Bearer', () => {
    return expect(() => AuthService.ensureValidAuthorizationBearer('foo')).toThrow(ValidationError)
  })

  test('should fail if the value is invalid', () => {
    return expect(() => AuthService.ensureValidAuthorizationBearer('Bearer foo')).toThrow(
      InvalidCredentialsError
    )
  })

  test('should pass if the value is loginTokenPayload', () => {
    return expect(() =>
      AuthService.ensureValidAuthorizationBearer(`Bearer ${validJWTToken}`)
    ).not.toThrow()
  })
})

describe('AuthService - serverToServerTokenAuth', () => {
  test('should fail if user not found', async () => {
    const authService: any = DIContainer.sharedContainer.authService
    authService.createUserSessionAndToken = jest.fn()
    const credentials: ServerToServerAuthCredentials = {
      connectUserID: 'invalid-connectId',
      deviceId: 'valid-deviceId',
    }
    await expect(authService.serverToServerTokenAuth(credentials)).rejects.toThrow(
      AccountNotFoundError
    )
  })

  test('should call createUserSessionAndToken', async () => {
    const authService: any = DIContainer.sharedContainer.authService
    authService.userRepository.getOne = jest.fn(() => validBody)
    authService.createUserSessionAndToken = jest.fn()
    const credentials: ServerToServerAuthCredentials = {
      connectUserID: 'invalid-connectId',
      deviceId: 'valid-deviceId',
    }
    await authService.serverToServerTokenAuth(credentials)
    expect(authService.createUserSessionAndToken).toHaveBeenCalled()
  })
})

describe('AuthService - createUserSessionAndToken', () => {
  test('should call createUserSessions', async () => {
    const authService: any = DIContainer.sharedContainer.authService
    authService.ensureValidUserStatus = jest.fn()
    authService.createUserSessions = jest.fn(async () => {
      return { userToken: { token: 'token' } }
    })

    await authService.createUserSessionAndToken(validUser, 'deviceId', false, true)
    expect(authService.createUserSessions).toHaveBeenCalled()
  })
})
