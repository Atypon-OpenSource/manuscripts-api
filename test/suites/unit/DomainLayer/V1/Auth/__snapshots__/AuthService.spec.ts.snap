// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService - Login should log user in and generates new user token 1`] = `
Object {
  "user": Object {
    "_id": "User|9f338224-b0d5-45aa-b02c-21c7e0c3c07a",
    "email": "valid-user@manuscriptsapp.com",
    "name": "Valid System User",
  },
}
`;

exports[`AuthService - Login should log user in and updates existing token 1`] = `
Object {
  "user": Object {
    "_id": "User|9f338224-b0d5-45aa-b02c-21c7e0c3c07a",
    "email": "valid-user@manuscriptsapp.com",
    "name": "Valid System User",
  },
}
`;
